includes/gateways/paypal/refunds.php

```diff
@@ -13,117 +13,161 @@
+			'content'     => __( 'Transaction not refunded in PayPal, as checkbox was not selected.', 'easy-digital-downloads' )
+		) );
+
 		return;
 	}
 
-	// If the payment has already been refunded, bail.
-	if ( $payment->get_meta( '_edd_paypal_refunded', true ) ) {
+	$refund = edd_get_order( $refund_id );
+	if ( empty( $refund->total ) ) {
 		return;
 	}
 
-	// Process the refund.
 	try {
-		refund_transaction( $payment );
+		refund_transaction( $order, $refund );
 	} catch ( \Exception $e ) {
-		edd_insert_payment_note( $payment->ID, sprintf(
-		/* Translators: %s - The error message */
-			__( 'Failed to refund transaction in PayPal. Error Message: %s', 'easy-digital-downloads' ),
+		edd_debug_log( sprintf(
+			'Failure when processing refund #%d. Message: %s',
+			$refund->id,
 			$e->getMessage()
+		), true );
+
+		edd_add_note( array(
+			'object_id'   => $order->id,
+			'object_type' => 'order',
+			'user_id'     => is_admin() ? get_current_user_id() : 0,
+			'content'     => sprintf(
+				/* Translators: %d - ID of the refund; %s - error message from PayPal */
+				__( 'Failure when processing PayPal refund #%d: %s', 'easy-digital-downloads' ),
+				$refund->id,
+				$e->getMessage()
+			)
 		) );
 	}
-}
-
-add_action( 'edd_pre_refund_payment', __NAMESPACE__ . '\maybe_refund_transaction', 999 );
+}, 10, 3 );
 
 /**
  * Refunds a transaction in PayPal.
  *
  * @link  https://developer.paypal.com/docs/api/payments/v2/#captures_refund
  *
- * @param \EDD_Payment $payment
+ * @param \EDD_Payment|Order $payment_or_order
+ * @param Order|null         $refund_object
  *
  * @since 2.11
  * @throws Authentication_Exception
  * @throws API_Exception
  * @throws \Exception
  */
-function refund_transaction( \EDD_Payment $payment ) {
-	$transaction_id = $payment->transaction_id;
+function refund_transaction( $payment_or_order, Order $refund_object = null ) {
+	/*
+	 * Internally we want to work with an Order object, but we also need
+	 * an EDD_Payment object for backwards compatibility in the hooks.
+	 */
+	$order = $payment = false;
+	if ( $payment_or_order instanceof Order ) {
+		$order = $payment_or_order;
+		$payment = edd_get_payment( $order->id );
+	} elseif ( $payment_or_order instanceof \EDD_Payment ) {
+		$payment = $payment_or_order;
+		$order   = edd_get_order( $payment->ID );
+	}
+
+	if ( empty( $order ) || ! $order instanceof Order ) {
+		return;
+	}
+
+	$transaction_id = $order->get_transaction_id();
 
 	if ( empty( $transaction_id ) ) {
 		throw new \Exception( __( 'Missing transaction ID.', 'easy-digital-downloads' ) );
 	}
 
-	$mode = ( 'live' === $payment->mode ) ? API::MODE_LIVE : API::MODE_SANDBOX;
+	$mode = ( 'live' === $order->mode ) ? API::MODE_LIVE : API::MODE_SANDBOX;
 
 	$api = new API( $mode );
 
-	$response = $api->make_request( 'v2/payments/captures/' . urlencode( $transaction_id ) . '/refund' );
+	$args = $refund_object instanceof Order ? array( 'invoice_id' => $refund_object->id ) : array();
```

Comment:
Just noting that this is the ID of the refund and **not** the ID of the order, because it appears PayPal expects this to be unique per refund. (Initially it was set to the original order ID and PayPal gave me this error:

```json
{"issue":"DUPLICATE_INVOICE_ID","description":"Invoice ID was previously used to refund a capture."}
```